### PR TITLE
*: fix lock in `CheckAndPutRegion`

### PR DESCRIPTION
**Pull request by rleungx**
Signed-off-by: Ryan Leung <rleungx@gmail.com>

### What problem does this PR solve?

Issue Number: Ref #5648. 

### What is changed and how does it work?

The problem is the same as the above issue and was introduced by #5615. This problem only exists in the master, no need to cherry-pick.

```commit-message
```

### Check List

Tests

- Unit test

### Release note

```release-note
None
```

### CODE REVIEW

File: server/core/region.go [hunk 1]
```diff
@@ -759,6 +759,7 @@ func (r *RegionsInfo) CheckAndPutRegion(region *RegionInfo) []*RegionInfo {
 	if err != nil {
 		log.Debug("region is stale", zap.Stringer("origin", origin.GetMeta()), errs.ZapError(err))
 		// return the state region to delete.
+		r.t.Unlock()
```
Comment (lhy1024) [inline server/core/region.go:762]: Do we need to cherry pick?

**Review by lhy1024 — APPROVED** [review]

File: server/core/region.go (hunk 1 above)
Comment (rleungx) [reply server/core/region.go:762]: No, you can check the PR description.

**Pull request merged**